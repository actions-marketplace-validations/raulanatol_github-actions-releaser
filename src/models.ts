export const BUG_LABELS: string[] = ['bug', 'Type: Bug'];
export const FEATURE_LABELS: string[] = ['enhancement', 'Type: Enhancement', 'feature', 'Type: Feature'];

export enum IssueType {
  FEATURE,
  BUG,
  OTHER,
}

export interface IssueToRelease {
  id: number;
  title: string;
  url: string;
  user: string;
  labels: string[];
  type: IssueType;
}

export interface ReleaseNotesIssuesText {
  bugs: string;
  features: string;
  others: string;
}

export interface GithubIssueLabel {
  name: string;
}
