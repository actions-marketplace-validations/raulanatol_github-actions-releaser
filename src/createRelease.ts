import * as core from '@actions/core';
import { Context } from '@actions/github/lib/context';

const toTagName = (ref: string): string => ref.replace('refs/tags/', '');

export const createRelease = async (github, context: Context, notes: string) => {
  const {
    ref,
    repo: { repo, owner },
  } = context;

  const tagName = toTagName(ref);

  const newRelease = await github.repos.createRelease({
    owner,
    repo,
    tag_name: tagName,
    name: `Release ${tagName}`,
    body: notes,
    draft: false,
    prerelease: false,
  });

  const {
    data: { id: releaseId, html_url: htmlUrl, upload_url: uploadUrl },
  } = newRelease;

  core.setOutput('id', releaseId);
  core.setOutput('html_url', htmlUrl);
  core.setOutput('upload_url', uploadUrl);
};
