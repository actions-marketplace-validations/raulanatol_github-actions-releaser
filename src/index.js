const core = require('@actions/core');
const { GitHub, context } = require('@actions/github');
const generateReleaseNotes = require('./releaseNotes');

async function run() {
  const releaseNotes = generateReleaseNotes();

  console.log('>>', releaseNotes);
}

module.exports = run;
