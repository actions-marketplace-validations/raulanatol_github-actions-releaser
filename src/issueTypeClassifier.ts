import { BUG_LABELS, BUG_TITLE_STARTS, DOCS_TITLE_STARTS, FEATURE_LABELS, FEATURE_TITLE_STARTS, GitHubIssue, GitHubIssueLabel, GitHubPullRequest, IssueType, TEST_TITLE_STARTS, TOOLS_TITLE_STARTS } from './models';

export const classifyIssue = (issue: GitHubIssue | GitHubPullRequest): IssueType => {
  const typeFromLabels = getIssueTypeFromLabels(issue);
  if (typeFromLabels) {
    return typeFromLabels;
  }

  const typeFromTitle = getIssueTypeFromTitle(issue);
  if (typeFromTitle) {
    return typeFromTitle;
  }

  return 'other';
};

const extractLabels = (labels: GitHubIssueLabel[]): string[] =>
  labels.map(({ name = '' }) =>
    name.toLowerCase());

const getIssueTypeFromLabels = (issue: GitHubIssue | GitHubPullRequest): IssueType | undefined => {
  const labels = extractLabels(issue.labels || []);

  for (const label of labels) {
    if (BUG_LABELS.includes(label)) {
      return 'bug';
    }

    if (FEATURE_LABELS.includes(label)) {
      return 'feature';
    }
  }
};

const getIssueTypeFromTitle = (issue: GitHubIssue | GitHubPullRequest): IssueType | undefined => {
  const issueTitle = issue.title.toLowerCase();

  for (const title of BUG_TITLE_STARTS) {
    if (issueTitle.startsWith(title)) {
      return 'bug';
    }
  }

  for (const title of FEATURE_TITLE_STARTS) {
    if (issueTitle.startsWith(title)) {
      return 'feature';
    }
  }

  for (const title of TEST_TITLE_STARTS) {
    if (issueTitle.startsWith(title)) {
      return 'test';
    }
  }

  for (const title of TOOLS_TITLE_STARTS) {
    if (issueTitle.startsWith(title)) {
      return 'tools';
    }
  }

  for (const title of DOCS_TITLE_STARTS) {
    if (issueTitle.startsWith(title)) {
      return 'docs';
    }
  }
};
