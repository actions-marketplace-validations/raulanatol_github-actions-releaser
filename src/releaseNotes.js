function generateReleaseNotes(github, context) {


  const { owner, repo } = context.repo;
  console.log('>>', context);
  console.log('>>', owner);
  console.log('>>', repo);


  return '# What\'s changed';


}

module.exports = generateReleaseNotes;
