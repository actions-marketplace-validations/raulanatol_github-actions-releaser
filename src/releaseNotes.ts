import { BUG_LABELS, FEATURE_LABELS, GithubIssueLabel, IssueToRelease, IssueType, ReleaseNotesIssuesText } from './models';

const issueToReleaseNoteText = (issue: IssueToRelease) => `- ${issue.title} ([#${issue.id}](${issue.url})) @${issue.user}`;

export const issuesToText = (issues: IssueToRelease[]): string => issues.map(issueToReleaseNoteText).join('\n');

const toOthersText = (issues: IssueToRelease[]) => (issues.length ? `\n🛠 Others\n--\n\n${issuesToText(issues)}\n` : '');

const toBugsText = (issues: IssueToRelease[]) => (issues.length ? `\n🐛 Bug Fixes\n--\n\n${issuesToText(issues)}\n` : '');

const toFeaturesText = (issues: IssueToRelease[]) => (issues.length ? `\n🚀 Features\n--\n\n${issuesToText(issues)}\n` : '');

export const toReleaseNoteText = ({ bugs, features, others }: ReleaseNotesIssuesText) => `# What's changed\n${bugs}${features}${others}`;

const extractLabels = (labels: GithubIssueLabel[] = []): string[] => labels.map((label) => label.name);

const getIssueType = (labels: string[] = []): IssueType => {
  for (const label of labels) {
    if (BUG_LABELS.includes(label)) {
      return IssueType.BUG;
    }

    if (FEATURE_LABELS.includes(label)) {
      return IssueType.FEATURE;
    }
  }
  return IssueType.OTHER;
};

const getClosedIssues = async (github, previousReleaseDate: string | undefined, repo: string, owner: string): Promise<IssueToRelease[]> => {
  const request: any = {
    owner,
    repo,
    state: 'closed'
  };

  if (previousReleaseDate) {
    request.since = previousReleaseDate;
  }

  const githubClosedIssues = await github.issues.listForRepo(request);

  return githubClosedIssues.data.map((issue) => {
    const labels = extractLabels(issue.labels);
    return {
      id: issue.number,
      title: issue.title,
      url: issue.html_url,
      user: issue.user.login,
      labels,
      type: getIssueType(labels)
    };
  });
};

export const getLatestReleaseDate = async (github, repo: string, owner: string): Promise<string | undefined> => {
  try {
    const lastRelease = await github.repos.getLatestRelease({ owner, repo });
    return lastRelease.data.published_at;
  } catch (e) {
    return undefined;
  }
};

export const toReleaseNotesIssues = (closedIssues: IssueToRelease[] = []): ReleaseNotesIssuesText => {
  const bugs: IssueToRelease[] = [];
  const features: IssueToRelease[] = [];
  const others: IssueToRelease[] = [];

  for (const issue of closedIssues) {
    if (issue.type === IssueType.BUG) {
      bugs.push(issue);
    } else if (issue.type === IssueType.FEATURE) {
      features.push(issue);
    } else {
      others.push(issue);
    }
  }

  return {
    others: toOthersText(others),
    bugs: toBugsText(bugs),
    features: toFeaturesText(features)
  };
};

export const issuesToReleaseNotes = (issues: IssueToRelease[]): string => {
  const releaseNotesIssuesText: ReleaseNotesIssuesText = toReleaseNotesIssues(issues);
  return toReleaseNoteText(releaseNotesIssuesText);
};

export const releaseNotes = async (github, repo: string, owner: string): Promise<string> => {
  const previousReleaseDate = await getLatestReleaseDate(github, repo, owner);
  const closedIssues = await getClosedIssues(github, previousReleaseDate, repo, owner);
  return issuesToReleaseNotes(closedIssues);
};
