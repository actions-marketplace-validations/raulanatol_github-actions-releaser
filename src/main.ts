import * as core from '@actions/core';
import { releaseNotes } from './releaseNotes';
import { context, getOctokit } from '@actions/github';
import { createRelease } from './createRelease';
import { GitHub } from '@actions/github/lib/utils';

const getGithubClient = () => {
  const githubToken = process.env.GITHUB_TOKEN;
  if (!githubToken) {
    throw new Error('Invalid GITHUB_TOKEN');
  }
  return getOctokit(githubToken);
};

const initialize = ({ repo, owner }) =>
  ({ github: getGithubClient(), repo, owner });

async function run(): Promise<void> {
  try {
    const init = initialize(context.repo);
    const github: InstanceType<typeof GitHub> = init.github;
    const notes = await releaseNotes(github, init.repo, init.owner);
    await createRelease(github, context, notes);

    core.debug(`Notes: ${notes}`);
  } catch (error) {
    core.setFailed(error.message);
  }
}

run();
