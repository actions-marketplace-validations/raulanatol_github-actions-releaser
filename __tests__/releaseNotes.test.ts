import { getLatestReleaseDate, issuesToReleaseNotes, toReleaseNotesIssues, toReleaseNoteText } from '../src/releaseNotes';
import { IssueToRelease } from '../src/models';

describe('ReleaseNotes', () => {
  describe('toReleaseNoteText', () => {
    test('should return a valid release notes when an empty list of bugs is provided', () => {
      expect(
        toReleaseNoteText({
          features: '',
          bugs: '',
          others: '',
          enhancements: ''
        })
      ).toMatchSnapshot();
    });
  });

  describe('toReleaseNotesIssues', () => {
    test('should return a valid group of issues', () => {
      const issues: IssueToRelease[] = [
        { id: 123, title: 'Bug 1', user: 'testUser1', url: 'https://github.com/issue/issue_1' },
        { id: 2845, title: 'Bug 2', user: 'testUser2', url: 'https://github.com/issue/issue_2' },
        { id: 234, title: 'Bug 3', user: 'testUser3', url: 'https://github.com/issue/issue_3' }
      ] as IssueToRelease[];
      expect(toReleaseNotesIssues(issues)).toMatchSnapshot();
    });
  });

  describe('issuesToReleaseNotes', () => {
    test('should generate a complex release notes', () => {
      const issues: IssueToRelease[] = [
        { id: 123, title: 'Bug 1', user: 'testUser1', url: 'https://github.com/issue/issue_1', type: 'bug' },
        { id: 2845, title: 'Bug 2', user: 'testUser2', url: 'https://github.com/issue/issue_2', type: 'bug' },
        { id: 2145, title: 'Feature 1', user: 'testUser23', url: 'https://github.com/issue/issue_3', type: 'feature' },
        { id: 234, title: 'Bug 3', user: 'testUser3', url: 'https://github.com/issue/issue_4', type: 'bug' },
        { id: 23214, title: 'Feature 21', user: 'testUser1', url: 'https://github.com/issue/issue_5', type: 'feature' },
        { id: 12, title: 'Other 1', user: 'testUser1', url: 'https://github.com/issue/issue_6', type: 'other' },
        { id: 1020, title: 'test: 1', user: 'testUser3', url: 'https://github.com/issue/issue_19', type: 'test' }
      ] as IssueToRelease[];
      expect(issuesToReleaseNotes(issues)).toMatchSnapshot();
    });

    test('should generate a complex release notes with no others', () => {
      const issues: IssueToRelease[] = [
        { id: 123, title: 'Bug 1', user: 'testUser1', url: 'https://github.com/issue/issue_1', type: 'bug' },
        { id: 2845, title: 'Bug 2', user: 'testUser2', url: 'https://github.com/issue/issue_2', type: 'bug' },
        { id: 2145, title: 'Feature 1', user: 'testUser23', url: 'https://github.com/issue/issue_3', type: 'feature' },
        { id: 234, title: 'Bug 3', user: 'testUser3', url: 'https://github.com/issue/issue_4', type: 'bug' },
        { id: 23214, title: 'Feature 21', user: 'testUser1', url: 'https://github.com/issue/issue_5', type: 'feature' }
      ] as IssueToRelease[];
      expect(issuesToReleaseNotes(issues)).toMatchSnapshot();
    });

    test('should generate a complex release notes with no features', () => {
      const issues: IssueToRelease[] = [
        { id: 123, title: 'Bug 1', user: 'testUser1', url: 'https://github.com/issue/issue_1', type: 'bug' },
        { id: 2845, title: 'Bug 2', user: 'testUser2', url: 'https://github.com/issue/issue_2', type: 'bug' },
        { id: 234, title: 'Bug 3', user: 'testUser3', url: 'https://github.com/issue/issue_4', type: 'bug' },
        { id: 12, title: 'Other 1', user: 'testUser1', url: 'https://github.com/issue/issue_6', type: 'other' }
      ] as IssueToRelease[];
      expect(issuesToReleaseNotes(issues)).toMatchSnapshot();
    });

    test('should generate a complex release notes with no bugs', () => {
      const issues: IssueToRelease[] = [
        { id: 2145, title: 'Feature 1', user: 'testUser23', url: 'https://github.com/issue/issue_3', type: 'feature' },
        { id: 23214, title: 'Feature 21', user: 'testUser1', url: 'https://github.com/issue/issue_5', type: 'feature' },
        { id: 12, title: 'Other 1', user: 'testUser1', url: 'https://github.com/issue/issue_6', type: 'other' }
      ] as IssueToRelease[];
      expect(issuesToReleaseNotes(issues)).toMatchSnapshot();
    });

    test('should generate an empty release notes when no closed issues is provided', () => {
      const issues: IssueToRelease[] = [];
      expect(issuesToReleaseNotes(issues)).toMatchSnapshot();
    });
  });

  describe('getLatestReleaseDate', () => {
    test('should return null when the repository does not have any release yet', async () => {
      const github = {
        repos: {
          getLatestRelease: () => Promise.reject('Not found')
        }
      } as any;
      const result = await getLatestReleaseDate(github, 'fakeRepo', 'fakeOwner');
      expect(result).toBeNull();
    });
  });
});
