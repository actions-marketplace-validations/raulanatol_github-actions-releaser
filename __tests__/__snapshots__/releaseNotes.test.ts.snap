// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReleaseNotes issuesToReleaseNotes should generate a complex release notes 1`] = `
"# What's changed

🐛 Bug Fixes
--

- Bug 1 ([#123](https://github.com/issue/issue_1)) @testUser1
- Bug 2 ([#2845](https://github.com/issue/issue_2)) @testUser2
- Bug 3 ([#234](https://github.com/issue/issue_4)) @testUser3

🚀 Features
--

- Feature 1 ([#2145](https://github.com/issue/issue_3)) @testUser23
- Feature 21 ([#23214](https://github.com/issue/issue_5)) @testUser1

💄 Enhancements
--

- test: 1 ([#1020](https://github.com/issue/issue_19)) @testUser3

🛠 Others
--

- Other 1 ([#12](https://github.com/issue/issue_6)) @testUser1
"
`;

exports[`ReleaseNotes issuesToReleaseNotes should generate a complex release notes with no bugs 1`] = `
"# What's changed

🚀 Features
--

- Feature 1 ([#2145](https://github.com/issue/issue_3)) @testUser23
- Feature 21 ([#23214](https://github.com/issue/issue_5)) @testUser1

🛠 Others
--

- Other 1 ([#12](https://github.com/issue/issue_6)) @testUser1
"
`;

exports[`ReleaseNotes issuesToReleaseNotes should generate a complex release notes with no features 1`] = `
"# What's changed

🐛 Bug Fixes
--

- Bug 1 ([#123](https://github.com/issue/issue_1)) @testUser1
- Bug 2 ([#2845](https://github.com/issue/issue_2)) @testUser2
- Bug 3 ([#234](https://github.com/issue/issue_4)) @testUser3

🛠 Others
--

- Other 1 ([#12](https://github.com/issue/issue_6)) @testUser1
"
`;

exports[`ReleaseNotes issuesToReleaseNotes should generate a complex release notes with no others 1`] = `
"# What's changed

🐛 Bug Fixes
--

- Bug 1 ([#123](https://github.com/issue/issue_1)) @testUser1
- Bug 2 ([#2845](https://github.com/issue/issue_2)) @testUser2
- Bug 3 ([#234](https://github.com/issue/issue_4)) @testUser3

🚀 Features
--

- Feature 1 ([#2145](https://github.com/issue/issue_3)) @testUser23
- Feature 21 ([#23214](https://github.com/issue/issue_5)) @testUser1
"
`;

exports[`ReleaseNotes issuesToReleaseNotes should generate an empty release notes when no closed issues is provided 1`] = `
"# What's changed
"
`;

exports[`ReleaseNotes toReleaseNoteText should return a valid release notes when an empty list of bugs is provided 1`] = `
"# What's changed
"
`;

exports[`ReleaseNotes toReleaseNotesIssues should return a valid group of issues 1`] = `
Object {
  "bugs": "",
  "enhancements": "",
  "features": "",
  "others": "
🛠 Others
--

- Bug 1 ([#123](https://github.com/issue/issue_1)) @testUser1
- Bug 2 ([#2845](https://github.com/issue/issue_2)) @testUser2
- Bug 3 ([#234](https://github.com/issue/issue_3)) @testUser3
",
}
`;
